File: 프로그래머스/3/118669. 등산코스 정하기/등산코스 정하기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <climits>
#include <queue>
using namespace std;



struct xy{
    int now;
    int now_t;
};
struct cmp{
    bool operator()(xy a, xy b){
        return a.now_t > b.now_t;
    }
};

struct Road{
    int next;
    int t;
};

int start[50001];
int mou[50001];

vector<Road> roads[50001];

int dist[50001];

void dijkstra(vector<int> s){
    priority_queue<xy,vector<xy>,cmp> pq;
    for(auto c : s){
        dist[c] = 0;
        pq.push({c,0});
    }
    
    while(!pq.empty()){
        xy x = pq.top();
        pq.pop();
        for(auto road : roads[x.now]){
            if(start[road.next] == 1) continue;
            int dis = max(road.t, x.now_t);
            if(dist[road.next] > dis){
                dist[road.next] = dis;
                if(mou[road.next] == 1) continue;
                pq.push({road.next,dist[road.next]});
            }
        }
        
    }
}

vector<int> solution(int n, vector<vector<int>> paths, vector<int> gates, vector<int> summits) {
    vector<int> answer;
    
    for(auto path : paths){
        roads[path[0]].push_back({path[1],path[2]});
        roads[path[1]].push_back({path[0],path[2]});
    }
    
    for(auto gate : gates)
        start[gate] = 1;
    for(auto summit : summits)
        mou[summit] = 1;
    
    int Min = INT_MAX;
    int mountain = 0;
    
    fill(dist, dist + n + 1, INT_MAX);
    dijkstra(gates);
    for(auto m : summits){
        if(dist[m] < Min){
            Min = dist[m];
            mountain = m;
        }
        else if(dist[m] == Min){
            mountain = min(mountain, m);
        }
    }
    
    answer.push_back(mountain);
    answer.push_back(Min);
    
    
    return answer;
}

File: 프로그래머스/3/118669. 등산코스 정하기/등산코스 정하기.java
import java.io.*;
import java.util.*;
class Solution {
    static ArrayList<Road>[] roads;
    static int[] mountain, start, dp;
    public int[] solution(int n, int[][] paths, int[] gates, int[] summits) {
        int m = summits.length;
        int k = gates.length;
        
        roads = new ArrayList[n + 1];
        dp = new int[n + 1];
        for(int i = 0; i <= n; i++) {
            roads[i] = new ArrayList<>();
            dp[i] = Integer.MAX_VALUE;
        }
        
        for(int i = 0; i < paths.length; i++){
            roads[ paths[i][0] ].add(new Road(paths[i][1], paths[i][2]));
            roads[ paths[i][1] ].add(new Road(paths[i][0], paths[i][2]));
        }
        
        mountain = new int[50001];
        for(int i = 0; i < m; i++) mountain[ summits[i] ] = 1;
        start = new int[k];
        for(int i = 0; i < k; i++) start[i] = gates[i];
        
        dijkstra();
        
        int[] answer = new int[2];
        answer[1] = Integer.MAX_VALUE;
        for(int i = 1; i <= n; i++){
            if(mountain[i] != 1) continue;
            if(answer[1] > dp[i]){
                answer[0] = i;
                answer[1] = dp[i];
            }
            
        }
        
        return answer;
    }
    void dijkstra(){
        PriorityQueue<Road> pq = new PriorityQueue<>((a, b) -> a.dis - b.dis);
        for(int s : start){
            pq.add(new Road(s, 0)); // 출발지 삽입
            dp[s] = 0;
        }
        
        while(!pq.isEmpty()){
            Road x = pq.poll();
            if (dp[x.x] < x.dis) continue;
            if(mountain[x.x] == 1) continue;
            
            for(Road road : roads[x.x]){
                int dis = Math.max(dp[x.x], road.dis);
                if(dp[road.x] > dis){
                    dp[road.x] = dis;
                    pq.add(new Road(road.x, dis));
                }
            }
        }
    }
    
    
    static class Road{
        int x, dis;
        Road(int a, int b){
            x = a; dis = b;
        }
    }
}